core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -577,14 +582,42 @@ private SnapshotInfo inProgressSnapshot(SnapshotsInProgress.Entry entry) {
                         shardSnapshotStatus.updateStage(IndexShardSnapshotStatus.Stage.FAILURE);
                         shardSnapshotStatus.failure(shardFailure.reason());
                         shardStatus.put(shardId, shardSnapshotStatus);
+                        hasFailedShards = true;
                     } else {
-                        IndexShardSnapshotStatus shardSnapshotStatus =
-                            repository.getShardSnapshotStatus(snapshotInfo.snapshotId(), snapshotInfo.version(), indexId, shardId);
-                        shardStatus.put(shardId, shardSnapshotStatus);
+                        nonFailedShards.put(shardId, indexId);
                     }
                 }
             }
         }
+        for (Map.Entry<ShardId, IndexId> shard : nonFailedShards.entrySet()) {
+            final ShardId shardId = shard.getKey();
+            final IndexId indexId = shard.getValue();
+            IndexShardSnapshotStatus shardSnapshotStatus;
+            try {
+                shardSnapshotStatus = repository.getShardSnapshotStatus(
+                    snapshotInfo.snapshotId(), snapshotInfo.version(), indexId, shardId);
+            } catch (IndexShardRestoreFailedException e) {
```

Comment:
I was wondering if this logic can be moved inside `repository.getShardSnapshotStatus()` instead?